examples/echo.rs

```diff
@@ -58,190 +61,171 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
-                friend_connection_c.add_friend(pk);
-            }
-            Ok(())
-        };
-
-        let lossy_future = async {
-            while lossy_rx.next().await.is_some() {
-                // ignore
-            }
-            Ok(())
-        };
-
-        // Add TCP relays
-        let tcp_connections_c = tcp_connections.clone();
-        let mut relays_futures: Vec<_> = TCP_RELAYS.iter()
-            .map(|&(pk, saddr)| {
-                // get PK bytes of the relay
-                let relay_pk_bytes: [u8; 32] = FromHex::from_hex(pk).unwrap();
-                // create PK from bytes
-                let relay_pk = PublicKey::from_slice(&relay_pk_bytes).unwrap();
-
-                tcp_connections_c.add_relay_global(saddr.parse().unwrap(), relay_pk)
-            })
-            .map(|future| future.map_err(Error::from).boxed())
-            .collect();
-
-        let mut futures = vec![
-            dht_server.run_socket(socket, rx, stats).map_err(Error::from).boxed(),
-            lan_discovery_sender.run().map_err(Error::from).boxed(),
-            tcp_connections.run().map_err(Error::from).boxed(),
-            onion_client.run().map_err(Error::from).boxed(),
-            net_crypto.run().map_err(Error::from).boxed(),
-            friend_connections.run().map_err(Error::from).boxed(),
-            net_crypto_tcp_future.boxed(),
-            tcp_incoming_future.boxed(),
-            lossless_future.boxed(),
-            lossy_future.boxed(),
-            friend_future.boxed(),
-        ];
-
-        futures.append(&mut relays_futures);
-
-        future::try_join_all(futures)
-            .map_err(|e| error!("Processing ended with error: {:?}", e))
-            .await
-            .unwrap();
+    let lossy_future = async {
+        while lossy_rx.next().await.is_some() {
+            // ignore
+        }
+        Result::<(), Error>::Ok(())
     };
 
-    let mut runtime = tokio::runtime::Runtime::new()?;
-    runtime.block_on(future);
+    // Add TCP relays
+    for &(pk, saddr) in TCP_RELAYS.iter() {
+        // get PK bytes of the relay
+        let relay_pk_bytes: [u8; 32] = FromHex::from_hex(pk).unwrap();
+        // create PK from bytes
+        let relay_pk = PublicKey::from_slice(&relay_pk_bytes).unwrap();
+
+        tcp_connections.add_relay_global(saddr.parse().unwrap(), relay_pk).await.map_err(Error::from)?;
+    }
+
+    futures::select!(
+        res = dht_server.run_socket(socket, rx, stats).fuse() => res.map_err(Error::from)?,
+        res = lan_discovery_sender.run().fuse() => res.map_err(Error::from)?,
+        res = tcp_connections.run().fuse() => res.map_err(Error::from)?,
+        res = onion_client.run().fuse() => res.map_err(Error::from)?,
+        res = net_crypto.run().fuse() => res.map_err(Error::from)?,
+        res = friend_connections.run().fuse() => res.map_err(Error::from)?,
+        res = net_crypto_tcp_future.fuse() => res?,
+        // res = tcp_incoming_future => res?,
```

Comment:
How so?

Reply:
It doesn't compile now with this future, but should be fine once net_crypto is changed to use async fn.

Reply:
In the current PR or in another?

Reply:
Current PR, otherwise it will break echo server :)